src/ui/common/src/components/pages/workflow/id/hook.ts

```diff
@@ -20,11 +28,58 @@ export type useWorkflowOutputs = {
   workflowDagResultWithLoadingStatus: WorkflowDagResultWithLoadingStatus;
 };
 
+// useWorkflowIds ensures we use the URL parameter as ground-truth for fetching
+// workflow, dag, and result IDs. It includes additional hooks to ensure
+// redux states are in-sync.
+// This hook should be used for all pages that need to access a single DAG (or DAG result)
+// data.
+export function useWorkflowIds(apiKey: string): useWorkflowIdsOutputs {
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+  const {
+    id: wfIdParam,
+    dagId: dagIdParam,
+    dagResultId: dagResultIdParam,
+  } = useParams();
+
+  const { data: dagResults } = useDagResultsGetQuery(
+    { apiKey, workflowId: wfIdParam },
+    { skip: !wfIdParam }
+  );
+
+  // Select the first availale dag result if ID is not provided.
+  const dagResult = dagResultIdParam
+    ? (dagResults ?? []).filter((r) => r.id === dagResultIdParam)[0]
+    : dagResults[0];
+
+  useEffect(() => {
+    if (dagResult !== undefined) {
+      dispatch(
+        initializeDagOrResultPageIfNotExists({
+          workflowId: wfIdParam,
+          dagId: dagResult.dag_id,
+          dagResultId: dagResult.id,
+        })
+      );
+
+      if (!dagResultIdParam) {
+        navigate(`result/${encodeURI(dagResult.id)}`, { replace: true });
+      }
+    }
+  }, [wfIdParam, dagResultIdParam, dagResult]);
+
+  return {
+    workflowId: wfIdParam,
+    dagId: dagResult?.dag_id ?? dagIdParam,
+    dagResultId: dagResult?.id,
```

Comment:
Should this be `dagResultId: dagResult?.id ?? dagResultIdParam,`?

Reply:
We actually need to refine the `dagId` part once the `DagsGet` route is ready. I'm using `dagResult?.id` without fallback to ensure the object exists (and thus a valid ID is given).